README.pod
=pod

=encoding UTF-8

=head1 NAME

MooseX::Singleton - Turn your Moose class into a singleton

=head1 VERSION

version 0.30

=head1 SYNOPSIS

    package MyApp;
    use MooseX::Singleton;

    has env => (
        is      => 'rw',
        isa     => 'HashRef[Str]',
        default => sub { \%ENV },
    );

    package main;

    delete MyApp->env->{PATH};
    my $instance = MyApp->instance;
    my $same = MyApp->instance;

=head1 DESCRIPTION

A singleton is a class that has only one instance in an application.
C<MooseX::Singleton> lets you easily upgrade (or downgrade, as it were) your
L<Moose> class to a singleton.

All you should need to do to transform your class is to change C<use Moose> to
C<use MooseX::Singleton>. This module uses metaclass roles to do its magic, so
it should cooperate with most other C<MooseX> modules.

=head1 METHODS

A singleton class will have the following additional methods:

=head2 Singleton->instance

This returns the singleton instance for the given package. This method does
I<not> accept any arguments. If the instance does not yet exist, it is created
with its defaults values. This means that if your singleton requires
arguments, calling C<instance> will die if the object has not already been
initialized.

=head2 Singleton->initialize(%args)

This method can be called I<only once per class>. It explicitly initializes
the singleton object with the given arguments.

=head2 Singleton->_clear_instance

This clears the existing singleton instance for the class. Obviously, this is
meant for use only inside the class itself.

=head2 Singleton->new

This method currently works like a hybrid of C<initialize> and
C<instance>. However, calling C<new> directly will probably be deprecated in a
future release. Instead, call C<initialize> or C<instance> as appropriate.

=for Pod::Coverage init_meta

=head1 SOME CODE STOLEN FROM

=for stopwords Anders

Anders Nor Berle E<lt>debolaz@gmail.comE<gt>

=head1 AND PATCHES FROM

Ricardo SIGNES E<lt>rjbs@cpan.orgE<gt>

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=MooseX-Singleton>
(or L<bug-MooseX-Singleton@rt.cpan.org|mailto:bug-MooseX-Singleton@rt.cpan.org>).

There is also a mailing list available for users of this distribution, at
L<http://lists.perl.org/list/moose.html>.

There is also an irc channel available for users of this distribution, at
L<C<#moose> on C<irc.perl.org>|irc://irc.perl.org/#moose>.

=head1 AUTHOR

Shawn M Moore <code@sartak.org>

=head1 CONTRIBUTORS

=for stopwords Dave Rolsky Karen Etheridge Ricardo SIGNES Kaare Rasmussen Anders Nor Berle Jonathan Rockway Hans Dieter Pearcey

=over 4

=item *

Dave Rolsky <autarch@urth.org>

=item *

Karen Etheridge <ether@cpan.org>

=item *

Ricardo SIGNES <rjbs@cpan.org>

=item *

Kaare Rasmussen <kaare@jasonic.dk>

=item *

Anders Nor Berle <berle@cpan.org>

=item *

Jonathan Rockway <jon@jrock.us>

=item *

Hans Dieter Pearcey <hdp@weftsoar.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Shawn M Moore.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for Perl extension MooseX-Singleton

{{$NEXT}}

0.30      2016-11-23 04:46:06Z
    - move repository to github Moose organization
    - properly declare test prereqs

0.29 2011-12-08
    - Fix warnings_once.t to not die if Test::NoWarnings isn't installed (Kaare)

0.28 2011-12-05
    - Fix RT Bug #46086, "name used only once" (Thanks chisel for the patch and
      the test)

0.27 2011-05-19
    - Fix 006-cooperative.t - this test requires MooseX::StrictConstructor
      0.16+ to pass with Moose 2.00+. Reported by Alexandr Ciornii. RT
      #67965. (Dave Rolsky)

0.26 2011-02-26
    - This module will now work with Moose 1.9900+, but it still works with
      older Moose as well. (Dave Rolsky)

    - The test suite now uses Test::Fatal instead of Test::Exception. (Karen
      Etheridge)

0.25 2010-08-22
    - Fixes for latest Moose. (Dave Rolsky)

0.24 2010-07-15
    - Require Moose 0.94+. (Dave Rolsky)

0.23 2010-07-15
    - Use modern Moose APIs to avoid warnings from the next Moose. (Dave
      Rolsky)

0.22 2009-12-25
    - The ->initialize method was both broken and undocumented. RT
      #51260. (Dave Rolsky)
    - Updated docs to encourage use of ->instance and ->initialize, and
      discourage calling ->new directly. Doing so may be deprecated in a
      future release. (Dave Rolsky)

0.21 2009-09-12
    - Various modules in this class were trying to load the old pre-0.20
      classes, which have all been renamed. Tests passed if you had 0.19 or
      earlier installed, but failed for fresh installations. (Dave Rolsky)

0.20 2009-09-11
    - Converted this extension to use roles rather than metaclass
      subclasses. This means it will cooperate with other extensions on CPAN,
      like MooseX::StrictConstructor, etc. (Dave Rolsky)

0.19 2009-07-09
    - Remove ambiguity about copyright holder (Sartak)

0.18 2009-06-21
    - Changes to work Moose 0.82 (Dave Rolsky)

0.17 2009-04-24
    - Add ClassName->_clear_instance. (hdp)

0.16 2009-04-24
    - Changes to keep constructor inlining working with latest Moose &
      Class::MOP. This module will still work with any Moose from 0.74
      on up.

0.15 2009-04-07
    - Fixed to work with Moose 0.73_01

    - When an object was implicitly constructed by calling
      ClassName->attribute, it skipped the BUILD and BUILDARGS defined
      for a class. Report and test from Josh in RT #42690. (Dave
      Rolsky)

0.14 2009-01-22
    - Converted to use new method generation helpers in the most
      recent Moose (0.65) and Class::MOP (Dave Rolsky)

0.13 2008-12-08
    - Fixed to work with Moose 0.63

0.12 2008-09-20
    - Fixed to work with Moose 0.58

0.11 2008-09-05
    - Fixed to work with Moose 0.57. Fixes RT #39013. Reported by Jon Swartz (Dave Rolsky)

0.10 2008-09-01
    - No code changes, just a stable release to go with Moose 0.56 release (Dave Rolsky)

0.09_02 2008-08-22
    - bumped the Moose dependency to 0.55_01 (Dave Rolsky)

0.09_01 2008-08-21
    - fixes to work with the latest dev release of Moose - 0.55_01 (Dave Rolsky)

0.09 2008-06-27
    - fix immutability to work with Moose 0.51 (Dave Rolsky)

0.08 2008-05-24
    - fix make_immutable keyword in test (Sartak)
    - immutable singleton objects did not call triggers in the constructor (Dave Rolsky)
    - fix immutability to actually work with modern Moose 0.41+ (Dave Rolsky)

0.07 2008-03-07
    - bump Moose dep to 0.37 because that's when we got make_immutable (Sartak)

0.06 2008-03-05
    - singleton objects were broken when made_immutable (Dave Rolsky)

0.05 2008-02-03
    - avoid re-BUILD-ing existing singleton objects (rjbs)

0.04 2008-01-27
    - exception when ->new called with args and instance already init'd (rjbs)
    - added ->initialize method to remove any ambiguity with ->new (rjbs)

0.03 2007-12-16
    - reimplementation as a metaclass (Sartak)
    - initial CPAN release

0.02 2007-12-16
    - instance and new are really the same, cleanup (Sartak)

0.01 2007-04-20
    - initial implementation as a role (Debolaz)


lib/MooseX/Singleton.pm
package MooseX::Singleton;

use Moose 1.10 ();
use Moose::Exporter;
use MooseX::Singleton::Role::Object;
use MooseX::Singleton::Role::Meta::Class;
use MooseX::Singleton::Role::Meta::Instance;

our $VERSION = '0.31';

Moose::Exporter->setup_import_methods( also => 'Moose' );

sub init_meta {
    shift;
    my %p = @_;

    Moose->init_meta(%p);

    my $caller = $p{for_class};

    Moose::Util::MetaRole::apply_metaroles(
        for             => $caller,
        class_metaroles => {
            class => ['MooseX::Singleton::Role::Meta::Class'],
            instance =>
                ['MooseX::Singleton::Role::Meta::Instance'],
            constructor =>
                ['MooseX::Singleton::Role::Meta::Method::Constructor'],
        },
    );

    Moose::Util::MetaRole::apply_base_class_roles(
        for_class => $caller,
        roles =>
            ['MooseX::Singleton::Role::Object'],
    );

    return $caller->meta();
}


1;

# ABSTRACT: Turn your Moose class into a singleton

__END__

=pod

=head1 SYNOPSIS

    package MyApp;
    use MooseX::Singleton;

    has env => (
        is      => 'rw',
        isa     => 'HashRef[Str]',
        default => sub { \%ENV },
    );

    package main;

    delete MyApp->env->{PATH};
    my $instance = MyApp->instance;
    my $same = MyApp->instance;

=head1 DESCRIPTION

A singleton is a class that has only one instance in an application.
C<MooseX::Singleton> lets you easily upgrade (or downgrade, as it were) your
L<Moose> class to a singleton.

All you should need to do to transform your class is to change C<use Moose> to
C<use MooseX::Singleton>. This module uses metaclass roles to do its magic, so
it should cooperate with most other C<MooseX> modules.

=head1 METHODS

A singleton class will have the following additional methods:

=head2 Singleton->instance

This returns the singleton instance for the given package. This method does
I<not> accept any arguments. If the instance does not yet exist, it is created
with its defaults values. This means that if your singleton requires
arguments, calling C<instance> will die if the object has not already been
initialized.

=head2 Singleton->initialize(%args)

This method can be called I<only once per class>. It explicitly initializes
the singleton object with the given arguments.

=head2 Singleton->_clear_instance

This clears the existing singleton instance for the class. Obviously, this is
meant for use only inside the class itself.

=head2 Singleton->new

This method currently works like a hybrid of C<initialize> and
C<instance>. However, calling C<new> directly will probably be deprecated in a
future release. Instead, call C<initialize> or C<instance> as appropriate.

=for Pod::Coverage init_meta

=head1 SOME CODE STOLEN FROM

=for stopwords Anders

Anders Nor Berle E<lt>debolaz@gmail.comE<gt>

=head1 AND PATCHES FROM

Ricardo SIGNES E<lt>rjbs@cpan.orgE<gt>

=cut



lib/MooseX/Singleton/Role/Object.pm
package MooseX::Singleton::Role::Object;
use Moose::Role;
use Carp qw( confess );

our $VERSION = '0.31';

sub instance { shift->new }

sub initialize {
    my ( $class, @args ) = @_;

    my $existing = $class->meta->existing_singleton;
    confess "Singleton is already initialized" if $existing;

    return $class->new(@args);
}

override new => sub {
    my ( $class, @args ) = @_;

    my $existing = $class->meta->existing_singleton;
    confess "Singleton is already initialized" if $existing and @args;

    # Otherwise BUILD will be called repeatedly on the existing instance.
    # -- rjbs, 2008-02-03
    return $existing if $existing and !@args;

    return super();
};

sub _clear_instance {
    my ($class) = @_;
    $class->meta->clear_singleton;
}

no Moose::Role;

1;

# ABSTRACT: Object class role for MooseX::Singleton

__END__

=pod

=head1 DESCRIPTION

=for Pod::Coverage *EVERYTHING*

This just adds C<instance> as a shortcut for C<new>.

=cut



lib/MooseX/Singleton/Role/Meta/Class.pm
package MooseX::Singleton::Role::Meta::Class;
use Moose::Role;
use MooseX::Singleton::Role::Meta::Instance;
use MooseX::Singleton::Role::Meta::Method::Constructor;

our $VERSION = '0.31';

sub existing_singleton {
    my ($class) = @_;
    my $pkg = $class->name;

    no strict 'refs';

    # create exactly one instance
    if ( defined ${"$pkg\::singleton"} ) {
        return ${"$pkg\::singleton"};
    }

    return;
}

sub clear_singleton {
    my ($class) = @_;
    my $pkg = $class->name;
    no strict 'refs';
    undef ${"$pkg\::singleton"};
}

override _construct_instance => sub {
    my ($class) = @_;

    # create exactly one instance
    my $existing = $class->existing_singleton;
    return $existing if $existing;

    my $pkg = $class->name;
    no strict 'refs';
    no warnings 'once';
    return ${"$pkg\::singleton"} = super;
};

if ( $Moose::VERSION >= 1.9900 ) {
    override _inline_params => sub {
        my $self = shift;

        return
            'my $existing = do {',
                'no strict "refs";',
                'no warnings "once";',
                '\${"$class\::singleton"};',
            '};',
            'return ${$existing} if ${$existing};',
            super();
    };

    override _inline_extra_init => sub {
        my $self = shift;

        return '${$existing} = $instance;';
    };
}

no Moose::Role;

1;

# ABSTRACT: Metaclass role for MooseX::Singleton

__END__

=pod

=for Pod::Coverage *EVERYTHING*

=head1 DESCRIPTION

This metaclass role makes sure that there is only ever one instance of an
object for a singleton class. The first call to C<construct_instance> is run
normally (and then cached). Subsequent calls will return the cached version.

=cut



lib/MooseX/Singleton/Role/Meta/Instance.pm
package MooseX::Singleton::Role::Meta::Instance;
use Moose::Role;
use Scalar::Util qw(weaken blessed);

our $VERSION = '0.31';

sub get_singleton_instance {
    my ( $self, $instance ) = @_;

    return $instance if blessed $instance;

    # optimization: it's really slow to go through new_object for every access
    # so return the singleton if we see it already exists, which it will every
    # single except the first.
    no strict 'refs';
    return ${"$instance\::singleton"} if defined ${"$instance\::singleton"};

    # We need to go through ->new in order to make sure BUILD and
    # BUILDARGS get called.
    return $instance->meta->name->new;
}

override clone_instance => sub {
    my ( $self, $instance ) = @_;
    $self->get_singleton_instance($instance);
};

override get_slot_value => sub {
    my ( $self, $instance, $slot_name ) = @_;
    $self->is_slot_initialized( $instance, $slot_name )
        ? $self->get_singleton_instance($instance)->{$slot_name}
        : undef;
};

override set_slot_value => sub {
    my ( $self, $instance, $slot_name, $value ) = @_;
    $self->get_singleton_instance($instance)->{$slot_name} = $value;
};

override deinitialize_slot => sub {
    my ( $self, $instance, $slot_name ) = @_;
    delete $self->get_singleton_instance($instance)->{$slot_name};
};

override is_slot_initialized => sub {
    my ( $self, $instance, $slot_name, $value ) = @_;
    exists $self->get_singleton_instance($instance)->{$slot_name} ? 1 : 0;
};

override weaken_slot_value => sub {
    my ( $self, $instance, $slot_name ) = @_;
    weaken $self->get_singleton_instance($instance)->{$slot_name};
};

override inline_slot_access => sub {
    my ( $self, $instance, $slot_name ) = @_;
    sprintf "%s->meta->instance_metaclass->get_singleton_instance(%s)->{%s}",
        $instance, $instance, $slot_name;
};

no Moose::Role;

1;

# ABSTRACT: Instance metaclass role for MooseX::Singleton

__END__

=pod

=head1 DESCRIPTION

=for Pod::Coverage *EVERYTHING*

This role overrides all object access so that it gets the appropriate
singleton instance for the class.

=cut



lib/MooseX/Singleton/Role/Meta/Method/Constructor.pm
package MooseX::Singleton::Role::Meta::Method::Constructor;
use Moose::Role;

our $VERSION = '0.31';

if ( $Moose::VERSION < 1.9900 ) {
    override _initialize_body => sub {
        my $self = shift;

        # TODO:
        # the %options should also include a both
        # a call 'initializer' and call 'SUPER::'
        # options, which should cover approx 90%
        # of the possible use cases (even if it
        # requires some adaption on the part of
        # the author, after all, nothing is free)
        my $source = 'sub {';
        $source .= "\n" . 'my $class = shift;';

        $source .= "\n"
            . 'my $existing = do { no strict "refs"; no warnings "once"; \${"$class\::singleton"}; };';
        $source .= "\n" . 'return ${$existing} if ${$existing};';

        $source .= "\n" . 'return $class->Moose::Object::new(@_)';
        $source
            .= "\n"
            . '    if $class ne \''
            . $self->associated_metaclass->name . '\';';

        $source .= $self->_generate_params( '$params', '$class' );
        $source .= $self->_generate_instance( '$instance', '$class' );
        $source .= $self->_generate_slot_initializers;

        $source .= ";\n" . $self->_generate_triggers();
        $source .= ";\n" . $self->_generate_BUILDALL();

        $source .= ";\n" . 'return ${$existing} = $instance';
        $source .= ";\n" . '}';
        warn $source if $self->options->{debug};

        my $attrs = $self->_attributes;

        my @type_constraints
            = map { $_->can('type_constraint') ? $_->type_constraint : undef }
            @$attrs;

        my @type_constraint_bodies
            = map { defined $_ ? $_->_compiled_type_constraint : undef; }
            @type_constraints;

        my $defaults = [map { $_->default } @$attrs];

        my ( $code, $e ) = $self->_compile_code(
            code        => $source,
            environment => {
                '$meta'                   => \$self,
                '$attrs'                  => \$attrs,
                '$defaults'               => \$defaults,
                '@type_constraints'       => \@type_constraints,
                '@type_constraint_bodies' => \@type_constraint_bodies,
            },
        );

        $self->throw_error(
            "Could not eval the constructor :\n\n$source\n\nbecause :\n\n$e",
            error => $e, data => $source )
            if $e;

        $self->{'body'} = $code;
    };
}

# Ideally we'd be setting this in the constructor, but the new() methods in
# what the parent classes are not well-factored.
#
# This is all a nasty hack, though. We need to fix Class::MOP::Inlined to
# allow constructor class roles to say "if the parent class has role X,
# inline".
override _expected_method_class => sub {
    my $self = shift;

    my $super_value = super();
    if ( $super_value eq 'Moose::Object' ) {
        for my $parent ( map { Class::MOP::class_of($_) }
            $self->associated_metaclass->superclasses ) {
            return $parent->name
                if $parent->is_anon_class
                    && grep { $_->name eq 'Moose::Object' }
                    map { Class::MOP::class_of($_) } $parent->superclasses;
        }
    }

    return $super_value;
};

no Moose::Role;

1;

# ABSTRACT: Constructor method role for MooseX::Singleton

__END__

=pod

=head1 DESCRIPTION

This role overrides the generated object C<new> method so that it returns the
singleton if it already exists.

=cut

